lib/modules/datasource/custom/formats/html.ts

```diff
@@ -0,0 +1,47 @@
+import is from '@sindresorhus/is';
+import { readLocalFile } from '../../../../util/fs';
+import { parse } from '../../../../util/html';
+import type { Http } from '../../../../util/http';
+import type { ReleaseResult } from '../../types';
+import type { CustomDatasourceFetcher } from './types';
+
+function extractLinks(content: string): ReleaseResult {
+  const body = parse(content);
+
+  // node-html-parser doesn't parse anything inside <pre>
+  // but, for example, nginx wraps directory listings in <pre>
+  const pres = body
+    .getElementsByTagName('pre')
+    .map((pre) => parse(pre.textContent));
+
+  const links = [body, ...pres].flatMap((e) => e.getElementsByTagName('a'));
+  const hrefs = links
+    .map((node) => node.getAttribute('href'))
+    .filter(is.truthy);
+
+  const releases = hrefs.map((href) => {
+    return {
+      version: href,
+    };
+  });
+
+  return { releases };
+}
+
+export class HtmlFetcher implements CustomDatasourceFetcher {
+  async fetch(http: Http, registryURL: string): Promise<unknown> {
+    const response = await http.get(registryURL, {
+      headers: {
+        Accept: 'text/html',
+      },
+    });
+
+    return extractLinks(response.body);
+  }
+
+  async readFile(registryURL: string): Promise<unknown> {
+    const fileContent = await readLocalFile(registryURL, 'utf8');
+
+    return extractLinks(fileContent!);
```

Comment:
you should handle null / undefined from `readLocalFile`

Reply:
Will I have to add a test for this case (null from `readLocalFile`) now?

Reply:
And, by the way, for `json` and `yaml` it's `return ...parse(fileContent!);`

https://github.com/renovatebot/renovate/blob/f08bad2579aa4845a461c31a3f149dddddfa352f/lib/modules/datasource/custom/formats/json.ts#L14

https://github.com/renovatebot/renovate/blob/f08bad2579aa4845a461c31a3f149dddddfa352f/lib/modules/datasource/custom/formats/yaml.ts#L16